cmake/recipes/external/gmp.cmake

```diff
@@ -3,12 +3,46 @@ if(TARGET gmp::gmp)
 endif()
 
 # Download precompiled .dll on Windows
-include(gmp_mpfr)
+if(WIN32)
+  include(gmp_mpfr)
+  # Find_package will look for our downloaded lib on Windows, and system-wide on Linux/macOS
+  find_package(GMP REQUIRED)
+else()
+  message(STATUS "Third-party: creating target 'gmp::gmp'")
 
-message(STATUS "Third-party: creating target 'gmp::gmp'")
+  set(gmp_INSTALL ${PROJECT_BINARY_DIR}/gmp-install)
+  set(gmp_LIB_DIR ${gmp_INSTALL}/lib)
+  set(gmp_LIBRARY ${gmp_LIB_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}gmp${CMAKE_STATIC_LIBRARY_SUFFIX})
+  set(gmp_INCLUDE_DIR ${gmp_INSTALL}/include)
+  include(ProcessorCount)
+  ProcessorCount(Ncpu)
+  include(ExternalProject)
+  ExternalProject_Add(gmp
+  URL  https://gmplib.org/download/gmp/gmp-6.2.1.tar.xz
+  URL_MD5 0b82665c4a92fd2ade7440c13fcaa42b
+  UPDATE_DISCONNECTED true  # need this to avoid constant rebuild
+  PATCH_COMMAND 
+    curl "https://gmplib.org/repo/gmp/raw-rev/5f32dbc41afc" "|" git apply -v
+  CONFIGURE_HANDLED_BY_BUILD ON  # avoid constant reconfigure
+  CONFIGURE_COMMAND 
+    ${PROJECT_BINARY_DIR}/gmp-prefix/src/gmp/configure 
+    --disable-debug --disable-dependency-tracking --enable-cxx --with-pic
+    --prefix=${gmp_INSTALL}
+    --disable-shared
+  BUILD_COMMAND make -j${Ncpu}
+  INSTALL_COMMAND make -j${Ncpu} install
+  TEST_COMMAND ""
+  BUILD_BYPRODUCTS ${gmp_LIBRARY}
+  )
+  ExternalProject_Get_Property(gmp SOURCE_DIR)
+  set(gmp_LIBRARIES ${gmp_LIBRARY})
+  add_library(gmp::gmp INTERFACE IMPORTED GLOBAL)
+  file(MAKE_DIRECTORY ${gmp_INCLUDE_DIR})  # avoid race condition
```

Comment:
You can still get race condition if you try to compile a file that includes `<gmp.h>` before the external project is downloaded.

Reply:
This is not for a race condition at `make` time, but rather at `cmake` time. The following line:
```
target_include_directories(gmp::gmp INTERFACE ${gmp_INCLUDE_DIR})
```
will fail if that include directory doesn't exist (it doesn't matter if the actual headers are there yet. just need the path to exist).

Reply:
(this comment `# avoid race condition` is from a copy-paste. I don't really agree that it's a race condition. It has nothing to do with threads AFAICT. It's rather just that the directory doesn't exist yet)